File: client/common/connection.go
package common

import (
    "net"
)
const BUFFER_SIZE = 1024

// Send writes the entire message to the TCP connection,
// appending a null terminator, handling short writes explicitly.
func send(conn net.Conn, message string) error {
	message += "\000"
	data := []byte(message)

	totalWritten := 0
	for totalWritten < len(data) {
		n, err := conn.Write(data[totalWritten:])
		if err != nil {
			return err
		}
		if n == 0 {
			return &net.OpError{Op: "write", Net: "tcp", Err: err}
		}
		totalWritten += n
	}

	return nil
}


// readUpToDelimiter reads from connection until the delimiter is found
func readUpToDelimiter(conn net.Conn, delim string) (string, error) {
    buf := make([]byte, 0, BUFFER_SIZE)
    tmp := make([]byte, 1)
    for {
        _, err := conn.Read(tmp)
        if err != nil {
            return "", err
        }
        if string(tmp) == delim {
            break
        }
        buf = append(buf, tmp[0])
    }
    return string(buf), nil
}

File: server/common/connection.py
BUFFER_SIZE = 1024

def send(connection, message):
    """
    Send a message ensuring full delivery, appending a null character at the end.
    """
    full_message = message + "\0"
    encoded = full_message.encode("utf-8")
    total_sent = 0

    while total_sent < len(encoded):
        sent = connection.send(encoded[total_sent:])
        if sent == 0:
            raise RuntimeError("Socket connection broken while sending")
        total_sent += sent


def read_up_to_delimiter(connection, delimiter):
    """
    Read from the connection until the delimiter is found, handling partial reads correctly.
    """
    buffer = bytearray()
    delimiter_bytes = delimiter.encode("utf-8")

    while True:
        data = connection.recv(BUFFER_SIZE)
        if not data:
            raise RuntimeError("Socket closed unexpectedly while reading")
        buffer.extend(data)

        pos = buffer.find(delimiter_bytes)
        if pos >= 0:
            return buffer[:pos].decode("utf-8")
